### ISSUE TITLE
[MNG-3379] Parallel resolution of artifacts

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#5483]
**[Donald J. Brown](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mrdon)** opened **[MNG-3379](https://issues.apache.org/jira/browse/MNG-3379?redirect=false)** and commented

Artifacts should be resolved in parallel, grouped by group id's to get around the lack of synchronization in the local repository.  The patch does the following:
* Use a ThreadPoolExecutor to parallelize artifact resolution, but takes care not to resolve multiple artifacts from the same group id simultaneously. (requires Java 5)
* Makes the http wagon the default instead of the poor performing http-client

Disadvantages:
* Requires Java 5, but the backport jars could be substituted pretty easily
* Breaks some plugins due to commons-logging being in the Maven uber jar (required by commons-httpclient), notably the apt plugin (maybe more should use the isolatedRealm setting?)
* Screws up the progress monitor as multiple threads are updating it

Advantages:
* Much faster when combined with the http wagon (WAGON-98).  I was seeing 40% improvement on some test builds.

---

**Affects:** 2.0.8

**Attachments:**
- [parallel-resolution.diff](https://issues.apache.org/jira/secure/attachment/12712571/parallel-resolution.diff) (_7.98 kB_)
- [parallel-resolution-2.diff](https://issues.apache.org/jira/secure/attachment/12712117/parallel-resolution-2.diff) (_9.05 kB_)
- [parallel-resolution-3.diff](https://issues.apache.org/jira/secure/attachment/12712374/parallel-resolution-3.diff) (_9.02 kB_)

**Sub-tasks:**
- [MNG-4025](https://issues.apache.org/jira/browse/MNG-4025) Prominently document opt-out setting for parallel artifact resolution for users

**Issue Links:**
- [MNG-3204](https://issues.apache.org/jira/browse/MNG-3204) Parallel dependency fetching
 (_**"is duplicated by"**_)
- [MNG-3739](https://issues.apache.org/jira/browse/MNG-3739) Perform multiple web requests simultaneously.
 (_**"is duplicated by"**_)
- [MNG-4179](https://issues.apache.org/jira/browse/MNG-4179) [regression] Artifact download hangs upon transfer failure

- [MNG-4432](https://issues.apache.org/jira/browse/MNG-4432) reimplement parallel artifacts download

- [MNG-3004](https://issues.apache.org/jira/browse/MNG-3004) Allow build lifecycle to execute projects in parallel
 (_**"is depended upon by"**_)

71 votes, 28 watchers

**Comment by jira-importer** [comment]
**[Chris Custine](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ccustine)** commented

I like the way this is going, but if I am not mistaken it appears that the patch ignores certain settings such as -Dmaven.repo.local and mirror settings in ~/.m2/settings.xml.

**Comment by jira-importer** [comment]
**[Donald J. Brown](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mrdon)** commented

How so?  The patch simply schedules the artifact resolution differently, but the same resolution code should work.

**Comment by jira-importer** [comment]
**[Donald J. Brown](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mrdon)** commented

Updated patch that fixes a few concurrency issues, adds support for bigger builds, updates all wagon deps to rc1-SNAPSHOT

**Comment by jira-importer** [comment]
**[Donald J. Brown](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mrdon)** commented

Updated patch (version 3) that works with Java 1.4 and eliminates problem with commons-logging (depends on WAGONHTTP-17)

**Comment by jira-importer** [comment]
**[Chris Custine](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ccustine)** commented

Hey Don, you are right, it isn't your patch that broke this.  I was testing this patch against 2.0.9-SNAPSHOT from SVN and got your changes intermingled with some recent updates from SVN which broke maven.repo.local setting.

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

Hey Chris - have you created a new issue for the problems you are experiencing?

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

Don, can you clarify the 3rd patch? It seems that you are using the normal wagon-http, not the standalone one from WAGONHTTP-17 - is this because you just substituted it in your local repo, or did you actually get it working with the standard dependency?

**Comment by jira-importer** [comment]
**[Donald J. Brown](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mrdon)** commented

I substituted it in my local repo.  In my latest -db release, I changed the version to be more clear.

**Comment by jira-importer** [comment]
**[Donald J. Brown](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mrdon)** commented

Try struts2-core - https://svn.apache.org/repos/asf/struts/struts2/trunk/core  It uses the APT plugin, which depends on commons-logging.  Another one I've heard of is the xdoclet plugin.

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

might want to adjust the shading mechanism and do that in the distribution rather than the wagonhttp-17 method

**Comment by jira-importer** [comment]
**[Donald J. Brown](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mrdon)** commented

No, you don't need to be worried about file-locking.  The retrieval algorithm only retrieves artifacts from different groups in parallel, but ones from the same group in serial, which is where you would need file-locking.

**Comment by jira-importer** [comment]
**[Jeremy Hanna](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jeromatron)** commented

Are there any plans to patch this into a version of maven 2?  I had hoped it would make it into 2.0.9.  I was really looking forward to the speed improvements as a result of parallel/asynchronous downloads of dependencies.

**Comment by jira-importer** [comment]
**[Donald J. Brown](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mrdon)** commented

I recreated the MNG-3379 branch but without the wagon changes.  I also improved the console monitor so it now properly handles multiple files being retrieved simultaneously.  What more needs to be done before the changes are moved into the 2.0.x branch?

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

the branch for this issue looks ok to me - though I see you've made additional changes on a labeled branch as well.

Is the MNG-3379 branch the complete list of changes you want to submit?

**Comment by jira-importer** [comment]
**[Donald J. Brown](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mrdon)** commented

Yes, the changes on the other branch mostly have to do with backporting the fix to the 2.0.9 tag due to Resource not having a proper hashCode implementation back then.

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

got it - thanks

**Comment by jira-importer** [comment]
**[Brian E. Fox (imported)](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brianf@infinity.nu)** commented

This issue passed the core ITs for 2.0.9 (and most for 2.0.10), i would like to see some more ITs testing the parallel downloads specifically and then we could consider it for .11

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

Brian - what kind of ITs do you envisage for this? Something that downloads a large number of artifacts from a file:// repository and check they all arrive safely?

**Comment by jira-importer** [comment]
**[Brian E. Fox (imported)](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brianf@infinity.nu)** commented

Brett, yes something that would exercise this code. I've been running this branch locally for a few days and haven't noticed any issues yet. Bumping it back for consideration in 2.0.11

**Comment by jira-importer** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

Hi Don,

(reposting my email to you here, since I never got a reply)

I was wondering if you could help us get this feature ready for introduction into Maven 2.1.0. We're putting together a release plan for it right now at:

http://docs.codehaus.org/display/MAVEN/Maven+2.1.0+Release+Plan

As you can see, things have changed WRT the versioning in Maven's future. The former 2.1 code - on trunk - has been renamed to 3.0-SNAPSHOT since it contains extensive changes, and we've decided to introduce some intermediary releases that will incorporate less risky new features such as parallel downloads that don't fit on the 2.0.x codeline appropriately. We're also going to be putting 2.0.x in end-of-life mode very soon.

I've heard good things about your patch for parallel downloads, but since I'm trying to push to make sure we have adequate documentation and tests for everything new that goes into Maven, we're still coming up short on this feature.

Can you provide us with a small design document just laying out the problem you solved, including scoping (for instance discussion about groupId limitations, and maybe any known issues on CI systems where multiple Maven processes may be sharing a single local repository with large numbers of parallel downloads happening simultaneously) and a short blurb describing your solution.

Also, we need to create some tests that we can use to guard against regressions in the future. Since you know this code the best, it'd be very helpful if you could help us design some test cases that might verify key pieces of the code to make sure nothing is broken (for instance, nothing that might introduce threading problems or file-locking issues).

If you can provide some of this information, it will allow us to put this feature into 2.1.0. Otherwise, I'm not sure whether we have the resource-hours available to grok the code, reconstruct the design decisions you took, and come up with adequate tests.

**Comment by jira-importer** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

Starting consolidation of all 2.1.0-M* versions into 2.1.0-M1, which I'll then rename into 2.1.0

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

even better, configure number of parallel threads, just set it to 1 to turn it off?

**Comment by jira-importer** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

We need to follow up on the ConcurrentModificationException that's happening intermittently under stress. The discussion is here:

http://www.nabble.com/Synchronization-issue-with-parallel-downloads-td22136382.html
